.github/actions/validate-deployment-settings/action.yml

```diff
@@ -0,0 +1,106 @@
+      # It then merges them together based on major version (Ex. [[{"0.20": "u298rer3", "0.20": "r98u349u"}], ...]).
+      # And then finally compares the hashes for each major version, and returns the major versions where they differ.
+      run: |
+        release_version_hashes=$(jq '.deployment.${{ inputs.target }}.Release.spack | to_entries' ${{ inputs.settings-path }})
+        prerelease_version_hashes=$(jq '.deployment.${{ inputs.target }}.Prerelease.spack | to_entries' ${{ inputs.settings-path }})
+
+        unequal_hashes_for_version=$(jq --compact-output --raw-output --null-input \
+          --argjson r "$release_version_hashes" \
+          --argjson p "$prerelease_version_hashes" \
+          '$r + $p
+          | [
+              group_by(.key)[]
+              | select(.[0] != null and .[1] != null and .[0].value != .[1].value)
+              | .[0].key
+            ]
+          | @sh'
+        )
+
+        if [[ "$unequal_hashes_for_version" != "" ]]; then
+          msg="Different spack hashes between Prerelease and Release for major versions: ${unequal_hashes_for_version}. These should be equal to ensure equivalent builds."
+          echo "::${{ inputs.error-level }}::$msg"
+          echo "msg=$msg" >> $GITHUB_OUTPUT
+        fi
+
+    - name: Verify Spack Commits Exist
+      id: spack-hash-all-exist
+      env:
+        GH_TOKEN: ${{ github.token }}
+        CHECKOUT_DIR: ./spack-metadata
+      shell: bash
+      # Verify that every commit referenced actually exists on the
+      # `releases/VERSION` branch in access-nri/spack, but don't
+      # bother checking out the code.
+      run: |
+        gh repo clone access-nri/spack ${{ env.CHECKOUT_DIR }} -- --no-checkout --bare --filter=blob:none
+
+        # Essentially, pull out all the spack 'version: hash' sections and iterate
+        jq -cr '.deployment.${{ inputs.target }}[].spack | to_entries[] | "\(.key) \(.value)"' ${{ inputs.settings-path }} | while read -ra line; do
+          version=${line[0]}
+          hash=${line[1]}
+          echo "Checking if $hash is in $version"
+          if ! git -C ${{ env.CHECKOUT_DIR }} merge-base --is-ancestor $hash releases/v$version; then
+            echo "::${{ inputs.error-level }}::Commit $hash does not exist on branch releases/v$version"
+            failed=true
+          fi
+        done
+
+        if [ -n $failed ]; then
+          msg="Some commits referenced do not exist in access-nri/spack. Check the workflow logs."
+          echo "::${{ inputs.error-level }}::$msg"
+          echo "msg=$msg" >> $GITHUB_OUTPUT
+        fi
+
+    - name: Warning/Error Collector
+      id: collector
+      shell: bash
+      run: echo "failures=${{ join(steps.*.outputs.msg) }}" >> $GITHUB_OUTPUT
+
+    - name: Action Fail from Error
+      if: inputs.error-level == 'error' && join(steps.*.outputs.msg) != ''
```

Comment:
Yep. In the case where we want to error if there is an issue (`inputs.error-level == 'error'`) but there is nothing wrong (`join(steps.*.outputs.msg) == ''`).

Reply:
But this logic won't trigger `exit 1` if `inputs.error-level == 'error' && join(steps.*.outputs.msg) != ''`. So we could have an error but if there is no message it won't exit. No?

Reply:
I'm not too sure I understand...`inputs.error-level` is to decide whether the messages are `notice`, `warning` or `error`-flavoured. We only want to `exit 1` if we treat the messages like `error`s, and there were messages (AKA something went wrong) generated in the earlier steps.

Reply:
Oh right. To my mind the logic for deciding to exit or not would be controlled by the `error-level`, and emitting a message would be a side-effect. If there is a message emit it, otherwise don't.

You've made it so the presence of a message indicates an some error state, and the `error-level` is what to do about it if an error message is set.

Correct?

Reply:
Correct!